nomad/node_endpoint.go

```diff
@@ -1179,8 +1205,8 @@ func (n *Node) UpdateAlloc(args *structs.AllocUpdateRequest, reply *structs.Gene
 	if node == nil {
 		return fmt.Errorf("node %s not found", nodeID)
 	}
-	if node.Status != structs.NodeStatusReady {
-		return fmt.Errorf("node %s is %s, not %s", nodeID, node.Status, structs.NodeStatusReady)
+	if node.UnresponsiveStatus() {
+		return fmt.Errorf("node %s is not allow to update allocs while in status %s", nodeID, node.Status)
```

Comment:
The `UpdateAlloc` method has the following comment in the docstring:

> // Clients must first register and heartbeat successfully before they are able
> // to call this method.

So now we're changing that so that we reject `UpdateStatus` changes (heartbeat) until after `UpdateAlloc` is called, right? Let's update the docstring to match. It also couldn't hurt to add a note to `UpdateStatus` about the same logic.

Reply:
The comment is still true. What this PR changes is that the status value in the `UpdateStatus` request may not be applied by the server. 

For example, if the node is `initializing` the heartbeats will send `Status: "ready"` but the server will not transition the node to `ready` until `UpdateAlloc` is called.

What this PR does change is that it allows `UpdateAlloc` to be called when the client is `initializing` as well to allow for the state transition. Previously it was only allowed when `ready`.

Reply:
Right, I understand that the comment is still strictly true, but it only partially describes the temporal dependency between the values being applied. The text we have is fine, but expanding it to describe that the status updates won't be applied until after `UpdateAlloc` is called too will make it more obvious what the dependencies are the next time someone comes thru here.

Reply:
Ohh I see now, you're talking about the comment for `UpdateStatus`, I misunderstood and thought you were talking about the comment `UpdateAlloc` 😅 

I updated the comment in https://github.com/hashicorp/nomad/pull/15808/commits/6e1f1d8c4604bcc06287b788fc056eb3fcc0fd2e to add a bit more details and also an ASCII version of the state diagram above, hope this makes it more clear 👍

Reply:
Ah we do need to update `UpdateAlloc` 😬 

Updated in https://github.com/hashicorp/nomad/pull/15808/commits/8202f0b592109d835c768eecc8151149b78609c9.